<!DOCTYPE html><html><head><meta charset="utf-8"><style>body {
  width: 45em;
  border: 1px solid #ddd;
  outline: 1300px solid #fff;
  margin: 16px auto;
}

body .markdown-body
{
  padding: 30px;
}

@font-face {
  font-family: fontawesome-mini;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAzUABAAAAAAFNgAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABbAAAABwAAAAcZMzaOEdERUYAAAGIAAAAHQAAACAAOQAET1MvMgAAAagAAAA+AAAAYHqhde9jbWFwAAAB6AAAAFIAAAFa4azkLWN2dCAAAAI8AAAAKAAAACgFgwioZnBnbQAAAmQAAAGxAAACZVO0L6dnYXNwAAAEGAAAAAgAAAAIAAAAEGdseWYAAAQgAAAFDgAACMz7eroHaGVhZAAACTAAAAAwAAAANgWEOEloaGVhAAAJYAAAAB0AAAAkDGEGa2htdHgAAAmAAAAAEwAAADBEgAAQbG9jYQAACZQAAAAaAAAAGgsICJBtYXhwAAAJsAAAACAAAAAgASgBD25hbWUAAAnQAAACZwAABOD4no+3cG9zdAAADDgAAABsAAAAmF+yXM9wcmVwAAAMpAAAAC4AAAAusPIrFAAAAAEAAAAAyYlvMQAAAADLVHQgAAAAAM/u9uZ4nGNgZGBg4ANiCQYQYGJgBEJuIGYB8xgABMMAPgAAAHicY2Bm42OcwMDKwMLSw2LMwMDQBqGZihmiwHycoKCyqJjB4YPDh4NsDP+BfNb3DIuAFCOSEgUGRgAKDgt4AAB4nGNgYGBmgGAZBkYGEAgB8hjBfBYGCyDNxcDBwMTA9MHhQ9SHrA8H//9nYACyQyFs/sP86/kX8HtB9UIBIxsDXICRCUgwMaACRoZhDwA3fxKSAAAAAAHyAHABJQB/AIEAdAFGAOsBIwC/ALgAxACGAGYAugBNACcA/wCIeJxdUbtOW0EQ3Q0PA4HE2CA52hSzmZDGe6EFCcTVjWJkO4XlCGk3cpGLcQEfQIFEDdqvGaChpEibBiEXSHxCPiESM2uIojQ7O7NzzpkzS8qRqnfpa89T5ySQwt0GzTb9Tki1swD3pOvrjYy0gwdabGb0ynX7/gsGm9GUO2oA5T1vKQ8ZTTuBWrSn/tH8Cob7/B/zOxi0NNP01DoJ6SEE5ptxS4PvGc26yw/6gtXhYjAwpJim4i4/plL+tzTnasuwtZHRvIMzEfnJNEBTa20Emv7UIdXzcRRLkMumsTaYmLL+JBPBhcl0VVO1zPjawV2ys+hggyrNgQfYw1Z5DB4ODyYU0rckyiwNEfZiq8QIEZMcCjnl3Mn+pED5SBLGvElKO+OGtQbGkdfAoDZPs/88m01tbx3C+FkcwXe/GUs6+MiG2hgRYjtiKYAJREJGVfmGGs+9LAbkUvvPQJSA5fGPf50ItO7YRDyXtXUOMVYIen7b3PLLirtWuc6LQndvqmqo0inN+17OvscDnh4Lw0FjwZvP+/5Kgfo8LK40aA4EQ3o3ev+iteqIq7wXPrIn07+xWgAAAAABAAH//wAPeJyFlctvG1UUh+/12DPN1B7P3JnYjj2Ox4/MuDHxJH5N3UdaEUQLqBIkfQQioJWQ6AMEQkIqsPGCPwA1otuWSmTBhjtps2ADWbJg3EpIXbGouqSbCraJw7kzNo2dRN1cnXN1ZvT7zuuiMEI7ncizyA0URofRBJpCdbQuIFShYY+GZRrxMDVtih5TwQPHtXDFFSIKoWIbuREBjLH27Ny4MsbVx+uOJThavebgVrNRLAiYx06rXsvhxLgWx9xpfHdrs/ekc2Pl2cpPCVEITQpwbj8VQhfXSq2m+Wxqaq2D73Kne5e3NjHqQNj3CRYlJlgUl/jRNP+2Gs2pNYRQiOnmUaQDqm30KqKiTTWPWjboxnTWpvgxjXo0KrtZXAHt7hwIz0YVcj88JnKlJKi3NPAwLyDwZudSmJSMMJFDYaOkaol6XtESx3Gt1VTytdZJ3DCLeaVhVnCBH1fycHTxFXwPX+l2e3d6H/TufGGmMTLTnbSJUdo00zuBswMO/nl3YLeL/wnu9/limCuD3vC54h5NBVz6Li414AI8Vx3iiosKcQXUbrvhFFiYb++HN4DaF4XzFW0fIN4XDWJ3a3XQoq9V8WiyRmdsatV9xUcHims1JloH0YUa090G3Tro3mC6c01f+YwCPquINr1PTaCP6rVTOOmf0GE2dBc7zWIhji3/5MchSuBHgDbU99RMWt3YUNMZMJmx92YP6NsHx/5/M1yvInpnkIOM3Z8fA3JQ2lW1RFC1KaBPDFXNAHYYvGy73aYZZZ3HifbeuiVZCpwA3oQBs0wGPYJbJfg60xrKEbKiNtTe1adwrpBRwlAuQ3q3VRaX0QmQ9a49BTSCuF1MLfQ6+tinOubRBZuWPNoMevGMT+V41KitO1is3D/tpMcq1JHZqDHGs8DoYGDkxJgKjHROeTCmhZvzPm9pod+ltKm4PN7Dyvvldlpsg8D+4AUJZ3F/JBstZz7cbFRxsaAGV6yX/dkcycWf8eS3QlQea+YLjdm3yrOnrhFpUyKVvFE4lpv4bO3Svx/6F/4xmiDu/RT5iI++lko18mY1oX+5UGKR6kmVjM/Zb76yfHtxy+h/SyQ0lLdpdKy/lWB6szatetQJ8nZ80A2Qt6ift6gJeavU3BO4gtxs/KCtNPVibCtYCWY3SIlSBPKXZALXiIR9oZeJ1AuMyxLpHIy/yO7vSiSE+kZvk0ihJ30HgHfzZtEMmvV58x6dtqns0XTAW7Vdm4HJ04OCp/crOO7rd9SGxQAE/mVA9xRN+kVSMRFF6S9JFGUtthkjBA5tFCWc2l4V43Ex9GmUP3SI37Jjmir9KqlaDJ4S4JB3vuM/jzyH1+8MuoZ+QGzfnvPoJb96cZlWjMcKLfgDwB7E634JTY+asjsPzS5CiVnEWY+KsrsIN5rn3mAPjqmQBxGjcGKB9f9ZxY3mYC2L85CJ2FXIxKKyHk+dg0FHbuEc7D5NzWUX32WxFcWNGRAbvwSx0RmIXVDuYySafluQBmzA/ssqJAMLnli+WIC90Gw4lm85wcp0qjArEDPJJV/sSx4P9ungTpgMw5gVC1XO4uULq0s3v1rqLi0vX/z65vlH50f8T/RHmSPTk5xxWBWOluMT6WiOy+tdvWxlV/XQb3o3c6Ssr+r6I708GsX9/nzp1tKFh0s3v7m4vAy/Hnb/KMOvc1wump6Il48K6mGDy02X9Yd65pa+nQIjk76lWxCkG8NBCP0HQS9IpAAAeJxjYGRgYGBhcCrq214Qz2/zlUGenQEEzr/77oug/zewFbB+AHI5GJhAogBwKQ0qeJxjYGRgYH3/P46BgZ0BBNgKGBgZUAEPAE/7At0AAAB4nGNngAB2IGYjhBsYBAAIYADVAAAAAAAAAAAAAFwAyAEeAaACCgKmAx4DggRmAAAAAQAAAAwAagAEAAAAAAACAAEAAgAWAAABAAChAAAAAHiclZI7bxQxFIWPd/JkUYQChEhIyAVKgdBMskm1QkKrRETpQiLRUczueB/K7HhlOxttg8LvoKPgP9DxFxANDR0tHRWi4NjrPIBEgh1p/dm+vufcawNYFWsQmP6e4jSyQB2fI9cwj++RE9wTjyPP4LYoI89iWbyLPIe6+Bh5Hs9rryMv4GbtW+RF3EhuRa7jbrIbeQkPkjdUETOLnL0Kip4FVvAhco1RXyMnSPEz8gzWxE7kWTwUp5HnsCLeR57HW/El8gJWa58iL+JO7UfkOh4l9yMv4UnyEtvQGGECgwF66MNBooF1bGCL1ELB/TYU+ZBRlvsKQ44Se6jQ4a7hef+fh72Crv25kp+8lNWGmeKoOI5jJLb1aGIGvb6TjfWNLdkqdFvJw4l1amjlXtXRZqRN7lSRylZZyhBqpVFWmTEXgWfUrpi/hZOQXdOd4rKuXOtEWT3k5IArPRzTUU5tHKjecZkTpnVbNOnt6jzN8240GD4xtikvZW56043rPMg/dS+dlOceXoR+WPbJ55Dsekq1lJpnypsMUsYOdCW30o103Ytu/lvh+5RWFLfBjm9/N8hJntPhvx92rnoE/kyHdGasGy754kw36vsVf/lFeBi+0COu+cfgQr42G3CRpeLoZ53gmfe3X6rcKt5oVxnptHR9JS8ehVUd5wvvahN2uqxOOpMXapibI5k7Zwbt4xBSaTfoKBufhAnO/uqNcfK8OTs0OQ6l7JIqFjDhYj5WcjevCnI/1DDiI8j4ndWb/5YzDZWh79yomWXeXj7Nnw70/2TIeFPTrlSh89k1ObOSRVZWZfgF0r/zJQB4nG2JUQuCQBCEd07TTg36fb2IyBaLd3vWaUh/vmSJnvpgmG8YcmS8X3Shf3R7QA4OBUocUKHGER5NNbOOEvwc1txnuWkTRb/aPjimJ5vXabI+3VfOiyS15UWvyezM2xiGOPyuMohOH8O8JiO4Af+FsAGNAEuwCFBYsQEBjlmxRgYrWCGwEFlLsBRSWCGwgFkdsAYrXFhZsBQrAAA=) format('woff');
}

@font-face {
  font-family: octicons-anchor;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAYcAA0AAAAACjQAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABMAAAABwAAAAca8vGTk9TLzIAAAFMAAAARAAAAFZG1VHVY21hcAAAAZAAAAA+AAABQgAP9AdjdnQgAAAB0AAAAAQAAAAEACICiGdhc3AAAAHUAAAACAAAAAj//wADZ2x5ZgAAAdwAAADRAAABEKyikaNoZWFkAAACsAAAAC0AAAA2AtXoA2hoZWEAAALgAAAAHAAAACQHngNFaG10eAAAAvwAAAAQAAAAEAwAACJsb2NhAAADDAAAAAoAAAAKALIAVG1heHAAAAMYAAAAHwAAACABEAB2bmFtZQAAAzgAAALBAAAFu3I9x/Nwb3N0AAAF/AAAAB0AAAAvaoFvbwAAAAEAAAAAzBdyYwAAAADP2IQvAAAAAM/bz7t4nGNgZGFgnMDAysDB1Ml0hoGBoR9CM75mMGLkYGBgYmBlZsAKAtJcUxgcPsR8iGF2+O/AEMPsznAYKMwIkgMA5REMOXicY2BgYGaAYBkGRgYQsAHyGMF8FgYFIM0ChED+h5j//yEk/3KoSgZGNgYYk4GRCUgwMaACRoZhDwCs7QgGAAAAIgKIAAAAAf//AAJ4nHWMMQrCQBBF/0zWrCCIKUQsTDCL2EXMohYGSSmorScInsRGL2DOYJe0Ntp7BK+gJ1BxF1stZvjz/v8DRghQzEc4kIgKwiAppcA9LtzKLSkdNhKFY3HF4lK69ExKslx7Xa+vPRVS43G98vG1DnkDMIBUgFN0MDXflU8tbaZOUkXUH0+U27RoRpOIyCKjbMCVejwypzJJG4jIwb43rfl6wbwanocrJm9XFYfskuVC5K/TPyczNU7b84CXcbxks1Un6H6tLH9vf2LRnn8Ax7A5WQAAAHicY2BkYGAA4teL1+yI57f5ysDNwgAC529f0kOmWRiYVgEpDgYmEA8AUzEKsQAAAHicY2BkYGB2+O/AEMPCAAJAkpEBFbAAADgKAe0EAAAiAAAAAAQAAAAEAAAAAAAAKgAqACoAiAAAeJxjYGRgYGBhsGFgYgABEMkFhAwM/xn0QAIAD6YBhwB4nI1Ty07cMBS9QwKlQapQW3VXySvEqDCZGbGaHULiIQ1FKgjWMxknMfLEke2A+IJu+wntrt/QbVf9gG75jK577Lg8K1qQPCfnnnt8fX1NRC/pmjrk/zprC+8D7tBy9DHgBXoWfQ44Av8t4Bj4Z8CLtBL9CniJluPXASf0Lm4CXqFX8Q84dOLnMB17N4c7tBo1AS/Qi+hTwBH4rwHHwN8DXqQ30XXAS7QaLwSc0Gn8NuAVWou/gFmnjLrEaEh9GmDdDGgL3B4JsrRPDU2hTOiMSuJUIdKQQayiAth69r6akSSFqIJuA19TrzCIaY8sIoxyrNIrL//pw7A2iMygkX5vDj+G+kuoLdX4GlGK/8Lnlz6/h9MpmoO9rafrz7ILXEHHaAx95s9lsI7AHNMBWEZHULnfAXwG9/ZqdzLI08iuwRloXE8kfhXYAvE23+23DU3t626rbs8/8adv+9DWknsHp3E17oCf+Z48rvEQNZ78paYM38qfk3v/u3l3u3GXN2Dmvmvpf1Srwk3pB/VSsp512bA/GG5i2WJ7wu430yQ5K3nFGiOqgtmSB5pJVSizwaacmUZzZhXLlZTq8qGGFY2YcSkqbth6aW1tRmlaCFs2016m5qn36SbJrqosG4uMV4aP2PHBmB3tjtmgN2izkGQyLWprekbIntJFing32a5rKWCN/SdSoga45EJykyQ7asZvHQ8PTm6cslIpwyeyjbVltNikc2HTR7YKh9LBl9DADC0U/jLcBZDKrMhUBfQBvXRzLtFtjU9eNHKin0x5InTqb8lNpfKv1s1xHzTXRqgKzek/mb7nB8RZTCDhGEX3kK/8Q75AmUM/eLkfA+0Hi908Kx4eNsMgudg5GLdRD7a84npi+YxNr5i5KIbW5izXas7cHXIMAau1OueZhfj+cOcP3P8MNIWLyYOBuxL6DRylJ4cAAAB4nGNgYoAALjDJyIAOWMCiTIxMLDmZedkABtIBygAAAA==) format('woff');
}

.markdown-body {
  font-family: sans-serif;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  color: #333333;
  overflow: hidden;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif;
  font-size: 16px;
  line-height: 1.6;
  word-wrap: break-word;
}

.markdown-body a {
  background: transparent;
}

.markdown-body a:active,
.markdown-body a:hover {
  outline: 0;
}

.markdown-body b,
.markdown-body strong {
  font-weight: bold;
}

.markdown-body mark {
  background: #ff0;
  color: #000;
  font-style: italic;
  font-weight: bold;
}

.markdown-body sub,
.markdown-body sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
.markdown-body sup {
  top: -0.5em;
}
.markdown-body sub {
  bottom: -0.25em;
}

.markdown-body h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

.markdown-body img {
  border: 0;
}

.markdown-body hr {
  -moz-box-sizing: content-box;
  box-sizing: content-box;
  height: 0;
}

.markdown-body pre {
  overflow: auto;
}

.markdown-body code,
.markdown-body kbd,
.markdown-body pre,
.markdown-body samp {
  font-family: monospace, monospace;
  font-size: 1em;
}

.markdown-body input {
  color: inherit;
  font: inherit;
  margin: 0;
}

.markdown-body html input[disabled] {
  cursor: default;
}

.markdown-body input {
  line-height: normal;
}

.markdown-body input[type="checkbox"] {
  box-sizing: border-box;
  padding: 0;
}

.markdown-body table {
  border-collapse: collapse;
  border-spacing: 0;
}

.markdown-body td,
.markdown-body th {
  padding: 0;
}

.markdown-body .codehilitetable {
  border: 0;
  border-spacing: 0;
}

.markdown-body .codehilitetable tr {
  border: 0;
}

.markdown-body .codehilitetable pre,
.markdown-body .codehilitetable div.codehilite {
  margin: 0;
}

.markdown-body .linenos,
.markdown-body .code,
.markdown-body .codehilitetable td {
  border: 0;
  padding: 0;
}

.markdown-body td:not(.linenos) .linenodiv {
  padding: 0 !important;
}

.markdown-body .code {
  width: 100%;
}

.markdown-body .linenos div pre,
.markdown-body .linenodiv pre,
.markdown-body .linenodiv {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-left-radius: 3px;
  -webkit-border-bottom-left-radius: 3px;
  -moz-border-radius-topleft: 3px;
  -moz-border-radius-bottomleft: 3px;
  border-top-left-radius: 3px;
  border-bottom-left-radius: 3px;
}

.markdown-body .code div pre,
.markdown-body .code div {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-right-radius: 3px;
  -webkit-border-bottom-right-radius: 3px;
  -moz-border-radius-topright: 3px;
  -moz-border-radius-bottomright: 3px;
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
}

.markdown-body * {
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body input {
  font: 13px Helvetica, arial, freesans, clean, sans-serif, "Segoe UI Emoji", "Segoe UI Symbol";
  line-height: 1.4;
}

.markdown-body a {
  color: #4183c4;
  text-decoration: none;
}

.markdown-body a:hover,
.markdown-body a:focus,
.markdown-body a:active {
  text-decoration: underline;
}

.markdown-body hr {
  height: 0;
  margin: 15px 0;
  overflow: hidden;
  background: transparent;
  border: 0;
  border-bottom: 1px solid #ddd;
}

.markdown-body hr:before,
.markdown-body hr:after {
  display: table;
  content: " ";
}

.markdown-body hr:after {
  clear: both;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 15px;
  margin-bottom: 15px;
  line-height: 1.1;
}

.markdown-body h1 {
  font-size: 30px;
}

.markdown-body h2 {
  font-size: 21px;
}

.markdown-body h3 {
  font-size: 16px;
}

.markdown-body h4 {
  font-size: 14px;
}

.markdown-body h5 {
  font-size: 12px;
}

.markdown-body h6 {
  font-size: 11px;
}

.markdown-body blockquote {
  margin: 0;
}

.markdown-body ul,
.markdown-body ol {
  padding: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body ol ol,
.markdown-body ul ol {
  list-style-type: lower-roman;
}

.markdown-body ul ul ol,
.markdown-body ul ol ol,
.markdown-body ol ul ol,
.markdown-body ol ol ol {
  list-style-type: lower-alpha;
}

.markdown-body dd {
  margin-left: 0;
}

.markdown-body code,
.markdown-body pre,
.markdown-body samp {
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
}

.markdown-body pre {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body kbd {
  background-color: #e7e7e7;
  background-image: -moz-linear-gradient(#fefefe, #e7e7e7);
  background-image: -webkit-linear-gradient(#fefefe, #e7e7e7);
  background-image: linear-gradient(#fefefe, #e7e7e7);
  background-repeat: repeat-x;
  border-radius: 2px;
  border: 1px solid #cfcfcf;
  color: #000;
  padding: 3px 5px;
  line-height: 10px;
  font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
  display: inline-block;
}

.markdown-body>*:first-child {
  margin-top: 0 !important;
}

.markdown-body>*:last-child {
  margin-bottom: 0 !important;
}

.markdown-body .headeranchor-link {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  display: block;
  padding-right: 6px;
  padding-left: 30px;
  margin-left: -30px;
}

.markdown-body .headeranchor-link:focus {
  outline: none;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  position: relative;
  margin-top: 1em;
  margin-bottom: 16px;
  font-weight: bold;
  line-height: 1.4;
}

.markdown-body h1 .headeranchor,
.markdown-body h2 .headeranchor,
.markdown-body h3 .headeranchor,
.markdown-body h4 .headeranchor,
.markdown-body h5 .headeranchor,
.markdown-body h6 .headeranchor {
  display: none;
  color: #000;
  vertical-align: middle;
}

.markdown-body h1:hover .headeranchor-link,
.markdown-body h2:hover .headeranchor-link,
.markdown-body h3:hover .headeranchor-link,
.markdown-body h4:hover .headeranchor-link,
.markdown-body h5:hover .headeranchor-link,
.markdown-body h6:hover .headeranchor-link {
  height: 1em;
  padding-left: 8px;
  margin-left: -30px;
  line-height: 1;
  text-decoration: none;
}

.markdown-body h1:hover .headeranchor-link .headeranchor,
.markdown-body h2:hover .headeranchor-link .headeranchor,
.markdown-body h3:hover .headeranchor-link .headeranchor,
.markdown-body h4:hover .headeranchor-link .headeranchor,
.markdown-body h5:hover .headeranchor-link .headeranchor,
.markdown-body h6:hover .headeranchor-link .headeranchor {
  display: inline-block;
}

.markdown-body h1 {
  padding-bottom: 0.3em;
  font-size: 2.25em;
  line-height: 1.2;
  border-bottom: 1px solid #eee;
}

.markdown-body h2 {
  padding-bottom: 0.3em;
  font-size: 1.75em;
  line-height: 1.225;
  border-bottom: 1px solid #eee;
}

.markdown-body h3 {
  font-size: 1.5em;
  line-height: 1.43;
}

.markdown-body h4 {
  font-size: 1.25em;
}

.markdown-body h5 {
  font-size: 1em;
}

.markdown-body h6 {
  font-size: 1em;
  color: #777;
}

.markdown-body p,
.markdown-body blockquote,
.markdown-body ul,
.markdown-body ol,
.markdown-body dl,
.markdown-body table,
.markdown-body pre,
.markdown-body .admonition {
  margin-top: 0;
  margin-bottom: 16px;
}

.markdown-body hr {
  height: 4px;
  padding: 0;
  margin: 16px 0;
  background-color: #e7e7e7;
  border: 0 none;
}

.markdown-body ul,
.markdown-body ol {
  padding-left: 2em;
}

.markdown-body ul ul,
.markdown-body ul ol,
.markdown-body ol ol,
.markdown-body ol ul {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body li>p {
  margin-top: 16px;
}

.markdown-body dl {
  padding: 0;
}

.markdown-body dl dt {
  padding: 0;
  margin-top: 16px;
  font-size: 1em;
  font-style: italic;
  font-weight: bold;
}

.markdown-body dl dd {
  padding: 0 16px;
  margin-bottom: 16px;
}

.markdown-body blockquote {
  padding: 0 15px;
  color: #777;
  border-left: 4px solid #ddd;
}

.markdown-body blockquote>:first-child {
  margin-top: 0;
}

.markdown-body blockquote>:last-child {
  margin-bottom: 0;
}

.markdown-body table {
  display: block;
  width: 100%;
  overflow: auto;
  word-break: normal;
  word-break: keep-all;
}

.markdown-body table th {
  font-weight: bold;
}

.markdown-body table th,
.markdown-body table td {
  padding: 6px 13px;
  border: 1px solid #ddd;
}

.markdown-body table tr {
  background-color: #fff;
  border-top: 1px solid #ccc;
}

.markdown-body table tr:nth-child(2n) {
  background-color: #f8f8f8;
}

.markdown-body img {
  max-width: 100%;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body code,
.markdown-body samp {
  padding: 0;
  padding-top: 0.2em;
  padding-bottom: 0.2em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(0,0,0,0.04);
  border-radius: 3px;
}

.markdown-body code:before,
.markdown-body code:after {
  letter-spacing: -0.2em;
  content: "\00a0";
}

.markdown-body pre>code {
  padding: 0;
  margin: 0;
  font-size: 100%;
  word-break: normal;
  white-space: pre;
  background: transparent;
  border: 0;
}

.markdown-body .codehilite {
  margin-bottom: 16px;
}

.markdown-body .codehilite pre,
.markdown-body pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f7f7f7;
  border-radius: 3px;
}

.markdown-body .codehilite pre {
  margin-bottom: 0;
  word-break: normal;
}

.markdown-body pre {
  word-wrap: normal;
}

.markdown-body pre code {
  display: inline;
  max-width: initial;
  padding: 0;
  margin: 0;
  overflow: initial;
  line-height: inherit;
  word-wrap: normal;
  background-color: transparent;
  border: 0;
}

.markdown-body pre code:before,
.markdown-body pre code:after {
  content: normal;
}

/* Admonition */
.markdown-body .admonition {
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  position: relative;
  border-radius: 3px;
  border: 1px solid #e0e0e0;
  border-left: 6px solid #333;
  padding: 10px 10px 10px 30px;
}

.markdown-body .admonition table {
  color: #333;
}

.markdown-body .admonition p {
  padding: 0;
}

.markdown-body .admonition-title {
  font-weight: bold;
  margin: 0;
}

.markdown-body .admonition>.admonition-title {
  color: #333;
}

.markdown-body .attention>.admonition-title {
  color: #a6d796;
}

.markdown-body .caution>.admonition-title {
  color: #d7a796;
}

.markdown-body .hint>.admonition-title {
  color: #96c6d7;
}

.markdown-body .danger>.admonition-title {
  color: #c25f77;
}

.markdown-body .question>.admonition-title {
  color: #96a6d7;
}

.markdown-body .note>.admonition-title {
  color: #d7c896;
}

.markdown-body .admonition:before,
.markdown-body .attention:before,
.markdown-body .caution:before,
.markdown-body .hint:before,
.markdown-body .danger:before,
.markdown-body .question:before,
.markdown-body .note:before {
  font: normal normal 16px fontawesome-mini;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  line-height: 1.5;
  color: #333;
  position: absolute;
  left: 0;
  top: 0;
  padding-top: 10px;
  padding-left: 10px;
}

.markdown-body .admonition:before {
  content: "\f056\00a0";
  color: 333;
}

.markdown-body .attention:before {
  content: "\f058\00a0";
  color: #a6d796;
}

.markdown-body .caution:before {
  content: "\f06a\00a0";
  color: #d7a796;
}

.markdown-body .hint:before {
  content: "\f05a\00a0";
  color: #96c6d7;
}

.markdown-body .danger:before {
  content: "\f057\00a0";
  color: #c25f77;
}

.markdown-body .question:before {
  content: "\f059\00a0";
  color: #96a6d7;
}

.markdown-body .note:before {
  content: "\f040\00a0";
  color: #d7c896;
}

.markdown-body .admonition::after {
  content: normal;
}

.markdown-body .attention {
  border-left: 6px solid #a6d796;
}

.markdown-body .caution {
  border-left: 6px solid #d7a796;
}

.markdown-body .hint {
  border-left: 6px solid #96c6d7;
}

.markdown-body .danger {
  border-left: 6px solid #c25f77;
}

.markdown-body .question {
  border-left: 6px solid #96a6d7;
}

.markdown-body .note {
  border-left: 6px solid #d7c896;
}

.markdown-body .admonition>*:first-child {
  margin-top: 0 !important;
}

.markdown-body .admonition>*:last-child {
  margin-bottom: 0 !important;
}

/* progress bar*/
.markdown-body .progress {
  display: block;
  width: 300px;
  margin: 10px 0;
  height: 24px;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #ededed;
  position: relative;
  box-shadow: inset -1px 1px 3px rgba(0, 0, 0, .1);
}

.markdown-body .progress-label {
  position: absolute;
  text-align: center;
  font-weight: bold;
  width: 100%; margin: 0;
  line-height: 24px;
  color: #333;
  text-shadow: 1px 1px 0 #fefefe, -1px -1px 0 #fefefe, -1px 1px 0 #fefefe, 1px -1px 0 #fefefe, 0 1px 0 #fefefe, 0 -1px 0 #fefefe, 1px 0 0 #fefefe, -1px 0 0 #fefefe, 1px 1px 2px #000;
  -webkit-font-smoothing: antialiased !important;
  white-space: nowrap;
  overflow: hidden;
}

.markdown-body .progress-bar {
  height: 24px;
  float: left;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #96c6d7;
  box-shadow: inset 0 1px 0 rgba(255, 255, 255, .5), inset 0 -1px 0 rgba(0, 0, 0, .1);
  background-size: 30px 30px;
  background-image: -webkit-linear-gradient(
    135deg, rgba(255, 255, 255, .4) 27%,
    transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%,
    transparent 77%, transparent
  );
  background-image: -moz-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -ms-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -o-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
}

.markdown-body .progress-100plus .progress-bar {
  background-color: #a6d796;
}

.markdown-body .progress-80plus .progress-bar {
  background-color: #c6d796;
}

.markdown-body .progress-60plus .progress-bar {
  background-color: #d7c896;
}

.markdown-body .progress-40plus .progress-bar {
  background-color: #d7a796;
}

.markdown-body .progress-20plus .progress-bar {
  background-color: #d796a6;
}

.markdown-body .progress-0plus .progress-bar {
  background-color: #c25f77;
}

.markdown-body .candystripe-animate .progress-bar{
  -webkit-animation: animate-stripes 3s linear infinite;
  -moz-animation: animate-stripes 3s linear infinite;
  animation: animate-stripes 3s linear infinite;
}

@-webkit-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@-moz-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

.markdown-body .gloss .progress-bar {
  box-shadow:
    inset 0 4px 12px rgba(255, 255, 255, .7),
    inset 0 -12px 0 rgba(0, 0, 0, .05);
}

/* Multimarkdown Critic Blocks */
.markdown-body .critic_mark {
  background: #ff0;
}

.markdown-body .critic_delete {
  color: #c82829;
  text-decoration: line-through;
}

.markdown-body .critic_insert {
  color: #718c00 ;
  text-decoration: underline;
}

.markdown-body .critic_comment {
  color: #8e908c;
  font-style: italic;
}

.markdown-body .headeranchor {
  font: normal normal 16px octicons-anchor;
  line-height: 1;
  display: inline-block;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.headeranchor:before {
  content: '\f05c';
}

.markdown-body .task-list-item {
  list-style-type: none;
}

.markdown-body .task-list-item+.task-list-item {
  margin-top: 3px;
}

.markdown-body .task-list-item input {
  margin: 0 4px 0.25em -20px;
  vertical-align: middle;
}

/* Media */
@media only screen and (min-width: 480px) {
  .markdown-body {
    font-size:14px;
  }
}

@media only screen and (min-width: 768px) {
  .markdown-body {
    font-size:16px;
  }
}

@media print {
  .markdown-body * {
    background: transparent !important;
    color: black !important;
    filter:none !important;
    -ms-filter: none !important;
  }

  .markdown-body {
    font-size:12pt;
    max-width:100%;
    outline:none;
    border: 0;
  }

  .markdown-body a,
  .markdown-body a:visited {
    text-decoration: underline;
  }

  .markdown-body .headeranchor-link {
    display: none;
  }

  .markdown-body a[href]:after {
    content: " (" attr(href) ")";
  }

  .markdown-body abbr[title]:after {
    content: " (" attr(title) ")";
  }

  .markdown-body .ir a:after,
  .markdown-body a[href^="javascript:"]:after,
  .markdown-body a[href^="#"]:after {
    content: "";
  }

  .markdown-body pre {
    white-space: pre;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  .markdown-body pre,
  .markdown-body blockquote {
    border: 1px solid #999;
    padding-right: 1em;
    page-break-inside: avoid;
  }

  .markdown-body .progress,
  .markdown-body .progress-bar {
    -moz-box-shadow: none;
    -webkit-box-shadow: none;
    box-shadow: none;
  }

  .markdown-body .progress {
    border: 1px solid #ddd;
  }

  .markdown-body .progress-bar {
    height: 22px;
    border-right: 1px solid #ddd;
  }

  .markdown-body tr,
  .markdown-body img {
    page-break-inside: avoid;
  }

  .markdown-body img {
    max-width: 100% !important;
  }

  .markdown-body p,
  .markdown-body h2,
  .markdown-body h3 {
    orphans: 3;
    widows: 3;
  }

  .markdown-body h2,
  .markdown-body h3 {
    page-break-after: avoid;
  }
}
</style><title>README</title></head><body><article class="markdown-body"><h5 id="projects-index"><a name="user-content-projects-index" href="#projects-index" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a><a href="/home/gire/Desktop/BiCro/Code/repo/00_gabriele/README.html">&#9664; Projects index</a></h5>
<style type="text/css">p + ul, p + ol { margin-top: -1em !important;  }</style>

<h1 id="genomic-loci-positioning-by-sequencing-gpseq-sequencing-data-analysis"><a name="user-content-genomic-loci-positioning-by-sequencing-gpseq-sequencing-data-analysis" href="#genomic-loci-positioning-by-sequencing-gpseq-sequencing-data-analysis" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Genomic loci Positioning by Sequencing (GPSeq) sequencing data analysis</h1>
<h2 id="the-main-script"><a name="user-content-the-main-script" href="#the-main-script" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>The <code>main</code> script</h2>
<pre><code>`main` runs the GPSeq sequencing data analysis pipeline.

Correct usage: `./main [-opt] settingsFile`

Required files:
    $DATA/BiCro-Data/&lt;expName&gt;/ folder containing *R1* [and *R2*] file[s]
    $DATA/BiCro-Data/&lt;expName&gt;/patfiles, a tab-separated file with condition|pattern|pattern_length
    $REPO/GPSeq/Sequencing/ folder containing the scripts

Options:
    `-h`  Shows this help page
    `-y`  Don't ask and perform every step.
    settingsFile path to a settings.sh file

Examples:
    `./main -h`
    `./main settings.sh`
    `./main -y settings.sh`
</code></pre>
<h4 id="pipeline-steps"><a name="user-content-pipeline-steps" href="#pipeline-steps" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Pipeline steps</h4>
<p>The steps performed by <code>main</code> are the following:</p>
<ol>
<li>Run <code>./quality_control</code>.</li>
<li>Run <code>./prepare_files</code></li>
<li>Perform pattern filtering:<ul>
<li>Count total reads and prepare <code>summary</code> file</li>
<li>Run <code>./pattern_filtering</code> per condition and update the <code>summary</code> file</li>
</ul>
</li>
<li>Perform read alignment.<ul>
<li>First, trim the pattern from the R1 reads (length specified in <code>pat_files</code>)</li>
<li>Run <code>./align_reads</code> and update summary with fraction of mapped reads</li>
<li>Add the patterns to the SAM files (condition.linker.sam)</li>
</ul>
</li>
<li>Filters the SAM file. (this step requires a specified <code>mapqthr</code> setting)<ul>
<li>Run <code>./sam_filter</code>.</li>
</ul>
</li>
<li>Prepare UMIs. If a <code>maskfile</code> was specified, UMIs are masked based on the information contained in such file.<ul>
<li>Run <code>./umi_group</code> to group UMIs on the same position.</li>
<li>Run <code>./pos2cutsite</code> to group UMIs on a cutsite</li>
<li>Run <code>./umi_dedupl</code> to deduplicate UMIs.</li>
</ul>
</li>
<li>Performs binning.<ul>
<li>Run <code>./cs_bin</code> to bin the cutsites.</li>
<li>Run <code>./umi_bin</code> to bin the UMIs.</li>
</ul>
</li>
<li>Analyzes UMIs.<ul>
<li>Run <code>./umi_plot</code> to generate analysis results as figures.</li>
</ul>
</li>
</ol>
<h4 id="output"><a name="user-content-output" href="#output" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Output</h4>
<p>Description of different output files.</p>
<h4 id="input"><a name="user-content-input" href="#input" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Input</h4>
<p>Two files are required: a <code>settings.sh</code> file used as input for the <code>main</code> script, and a <code>pat_files</code> file in the input directory, which contains the pattern to recognize the different conditions. Please, have a look at the following sample <code>settings.sh</code> file.</p>
<pre><code># -------- #
# SETTINGS #
# -------- #

# GENERAL
# ------------------------------
experiment='TK20'               # Experiment ID
conds='neg,1min,10min,2h'       # Comma-separated conditions list
neg='neg'                       # Specify negative control label, if present
numbproc=30                     # Number of threads to use

# ALIGNMENT
# ------------------------------
genome='hg19'                   # Reference genome
aligner='bwa'                   # Aligner: 'bwa' or 'bowtie2'
# SAM filtering

# ------------------------------
cutsite='AAGCTT'                # Cutsite
mapqthr=30                      # MAPQ threshold for SAM filtering
platform='L'                    # Sequencing platform used
                                #   S - Sanger, Phred+33
                                #   X - Solexa, Solexa+64
                                #   I - Illumina 1.3+, Phred+64
                                #   J - Illumina 1.5+, Phred+64
                                #   L - Illumina 1.8+, Phred+33
rmX=false                       # Remove X chromosome data
rmY=true                        # Remove Y chromosome data

# UMI analysis
# ------------------------------
umi_length=8                    # UMI length in nt
binsize=1e6                     # Window size for UMI analysis
binstep=1e5                     # Distance from consecutive windows
csrange=40                      # Window around cutsites (if cutsitelist is provided)

# Automatic filtering based on read identity probability
pthr=0                          # Threshold for UMI identity probability
                                # Set pthr=1 to use the automatic cutoff
                                # Set pthr=0 to skip automatic filtering
                                # The cutoff used is the smaller between pthr and the automatic one.

# Automatic filtering based on read quality
emax=1e-3                       # Maximum error probability for filtering.
eperc=20                        # Maximum percentage of bases with emax error probability.

# FILE PATHS
# ------------------------------
# P.s.: If you don't want to provide a file just assign a non existent path (e.g., '-')
#       If you assign an empty string (i.e., '') then a user input might be required

# Cutsite list file path, required columns: |chr|pos|
cutsitelist='/media/gire/Data/BiCro-Resources/hg19.HindIII.txt'

# Maskfile path, required columns: |id|chr|start|end|
maskfile='/media/gire/Data/BiCro-Data/Sequencing/TK20/consensusBlacklist.centro.telo.chr5Peak.tsv'

# Chromosome lengths file path, required columns: |chr|len| (no header)
chrlengths='/media/gire/Data/BiCro-Resources/hg19.chr_size.txt'
</code></pre>
<p>The <code>pat_files</code> instead, which is expected to be located in the <code>$DATA/BiCro-Data/Sequencing/$experiment</code> folder, contains a row per condition (as specified in the <code>settings.sh</code> file). Every row contains the following tab-separated information: condition label, scan_for_matches pattern, pattern length. An example being the following:</p>
<pre><code>neg ^ 8...8 CATCAGAA AAGCTT 1...1000 $  22
1min    ^ 8...8 CATCATCC AAGCTT 1...1000 $  22
10min   ^ 8...8 GTCGTTCC AAGCTT 1...1000 $  22
2h  ^ 8...8 TGATGTCC AAGCTT 1...1000 $  22
</code></pre>
<h2 id="other-scripts"><a name="user-content-other-scripts" href="#other-scripts" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Other scripts</h2>
<h3 id="quality_control"><a name="user-content-quality_control" href="#quality_control" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./quality_control</h3>
<p>This script is run by <code>main</code> during step 1. It performs a quality control with the <code>fastqc</code> tool.</p>
<pre><code>Correct usage: `./quality_control numbOfiles ncores outFolder r1file [r2file]`

Options:
    numbOfiles  1 for single-ended, 2 for pair-ended
    ncores      number of threads to use (-t option of fastqc)
    outFolder   output folder
    r1file      R1 fastq file
    [r2file]    R2 fastq file (optional, needed only if numbOfiles=2)
</code></pre>
<h3 id="prepare_files"><a name="user-content-prepare_files" href="#prepare_files" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./prepare_files</h3>
<p>This script is run by <code>main</code> during step 2. It decompresses the sequencing data and converts the generated files in different &lsquo;formats&rsquo;.</p>
<pre><code>Correct usage: `./prepare_files numbOfiles ncores outFolder r1file [r2file]`

Options:
    numbOfiles  1 for single-ended, 2 for pair-ended
    ncores      number of threads to use (-t option of fastqc)
    outputDir   output folder
    r1file      R1 compressed file
    [r2file]    R2 compressed file (optional, needed only for paired-end sequencing)
</code></pre>
<p>The produced files are the following:</p>
<ul>
<li><code>r1.fq</code> (and <code>r2.fq</code>): decompressed <strong>fastq</strong> file(s).</li>
<li><code>r2oneline.fq</code> (and <code>r2oneline.fq</code>): one read per line <strong>fastq</strong> file(s).</li>
<li><code>r1.fa</code> (and <code>r2.fa</code>): decompressed <strong>fasta</strong> file(s).</li>
<li><code>r1oneline.fa</code> (and <code>r2oneline.fa</code>): one read per line <strong>fasta</strong> file(s).</li>
</ul>
<h3 id="pattern_filtering"><a name="user-content-pattern_filtering" href="#pattern_filtering" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./pattern_filtering</h3>
<p>This script is run by <code>main</code> during step 3. It runs the <code>scan_for_matches</code> tool to select reads based on a pattern (e.g., containing barcodes), and divide them into different conditions.</p>
<pre><code>Correct usage: `./pattern_filtering inDir outCtrlDir outDir patFile nCores nFiles [cs]`

Options:
    nFiles      1 for single-ended, 2 for pair-ended
    inDir       outputDir of prepare_files script (with fa and fq files)
    outCtrlDir  temporary output folder
    outDir      output folder (usually, condition specific)
    patFile     path to pattern file
    ncores      number of threads to use
</code></pre>
<p>The script will create a <code>filtered.r1.fa</code> and a <code>filtered.r1.fq</code> (and respective r2 files in case of paired-end sequencing) in the specified <code>outDir</code>. These files will contain the reads that match the specified pattern, retrieved from <code>r1.fa</code> and <code>r1.fq</code> (and r2&hellip;).</p>
<p>Finally, the <code>summary</code> file is updated with the number and percentage of reads belonging to that condition.</p>
<h3 id="align_reads"><a name="user-content-align_reads" href="#align_reads" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./align_reads</h3>
<p>This BASH script is run by <code>main</code> in step 4. It aligns the reads to a reference genome using either <code>bwa</code> or <code>bowtie2</code> and generates <code>bam</code> and <code>bai</code> files.</p>
<pre><code>Correct usage: `./align_reads outDir condition genome ncores nfiles [aln]`

Options:
    outDir      condition directory (with previous steps output)
    condition   condition label
    genome      reference genome [e.g., 'hg19']
    ncores      number of threads to use
    nFiles      1 for single-ended, 2 for pair-ended
    [aln]       either 'bwa' or 'bowtie2', defaults to 'bwa'
</code></pre>
<h3 id="sam_filter"><a name="user-content-sam_filter" href="#sam_filter" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./sam_filter</h3>
<p>This R script is run by <code>main</code> in step 5. It filters a SAM file based on a variety of settings.</p>
<pre><code>usage:  umi_dedupl [--] [--help] [--opts OPTS] [--cutsite CUTSITE]
        [--mapq-thr MAPQ-THR] [--num-proc NUM-PROC] [--suffix SUFFIX]
        dirpath experiment condition

Deduplicate UMIs.

positional arguments:
  dirpath           Experiment ccondition directory, contains UMI counts.
  experiment        Experiment ID (e.g., TK26).
  condition         Condition folder name (e.g., 400U2h).

flags:
  -h, --help        show this help message and exit

optional arguments:
  -x, --opts OPTS               RDS file containing argument values
  -cs, --cutsite CUTSITE        Cutsite sequence (e.g., for HindIII AAGCTT).
                                [default: AAGCTT]
  -mt, --mapq-thr MAPQ-THR      Mapping quality threshold.
                                [default: 30]
  -c, --num-proc NUM-PROC       Number of cores for parallel computation.
                                [default: 1]
  -s, --suffix SUFFIX           Suffix to be added to output files.
                                [default: '']
</code></pre>
<p>Specifically, the filters do the following:</p>
<ul>
<li>Remove secondary alignments (based on SAM file flags).</li>
<li>If paired-end sequencing, remove chimeras (where R1 and R2 are mapped on different chromosomes).</li>
<li>Remove unmapped reads.</li>
<li>If paired-end sequencing, remove R2 and keep only R1.</li>
<li>Remove reads with a mapping quality lower than <code>mapqthr</code>.</li>
</ul>
<p>Then, the generated output consists of two files:</p>
<ul>
<li><code>condition.filtered.sam.RData</code> contains the raw filtered SAM file as an RData structure.</li>
<li><code>condition.filtered.umi.pos.txt</code> is a table with the following columns: chromosome, 3&rsquo;position, linker sequence, linker quality, read name. The position does not correspond to the SAM file one, but is instead shifted accordingly to the CIGAR string in the SAM file or to the length of the cutsite.</li>
</ul>
<h3 id="umi_group"><a name="user-content-umi_group" href="#umi_group" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./umi_group</h3>
<p>This Python script is run by <code>main</code> in step 6. It groups reads mapped on the same genomic locus.</p>
<pre><code>usage: umi_group [-h] [--mask-file MF] dirpath expID cond umiLen

Group UMIs based on their location.

positional arguments:
  dirpath         Experiment condition directory, contains UMI counts.
  expID           Experiment ID (e.g., TK26).
  cond            Condition folder name (e.g., 400U2h).
  umiLen          Length of the UMIs in nt (e.g., 8).

optional arguments:
  -h, --help      show this help message and exit
  --mask-file MF  File with regions to be masked.
</code></pre>
<p>The output consists of:</p>
<ul>
<li><code>UMIpos.txt</code>, containing the following columns:<ul>
<li>chromosome</li>
<li>position</li>
<li>space-separated UMIs</li>
<li>quality strings</li>
</ul>
</li>
<li>If a <code>maskfile</code> was specified, then an additional <code>UMIpos.all.txt</code> file will be generated, containing all the UMIs (also those withing masked regions).</li>
</ul>
<h3 id="pos2cutsite"><a name="user-content-pos2cutsite" href="#pos2cutsite" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./pos2cutsite</h3>
<p>This R script is run by <code>main</code> in step 6. It groups reads mapped on the same cutsite. Every read is assigned to the closest cutsite. Then, a read is discarded and considered &lsquo;orphan&rsquo;, if the closest cutsite is more than bin_size/2 nt away from the aligned locus. In other words, a window of bin_size nt is built around every cutsite.</p>
<p>The <code>pos2cutsite</code> step is not mandatory and, if the list of custite in the current reference genome is not known, the analysis can be performed anyway.</p>
<pre><code>usage:  umi_dedupl [--] [--help] [--opts OPTS] [--bin-size BIN-SIZE]
        [--num-proc NUM-PROC] [--suffix SUFFIX]
        dirpath experiment condition cutsites

Deduplicate UMIs.

positional arguments:
  dirpath           Experiment ccondition directory, contains UMI counts.
  experiment        Experiment ID (e.g., TK26).
  condition         Condition folder name (e.g., 400U2h).
  cutsites          File containing the cutsite positions. (chr | pos)

flags:
  -h, --help        show this help message and exit

optional arguments:
  -x, --opts OPTS           RDS file containing argument values
  -i, --bin-size BIN-SIZE   Range size in bp around the cutsite.
                            [default: 40]
  -c, --num-proc NUM-PROC   Number of cores for parallel computation.
                            [default: 1]
  -s, --suffix SUFFIX       Suffix to be added to output files.
                            [default: '']
</code></pre>
<h3 id="umi_dedupl"><a name="user-content-umi_dedupl" href="#umi_dedupl" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./umi_dedupl</h3>
<p>This R script is run by <code>main</code> in step 6. It removes duplicated (through amplification) reads based on the UMIs.</p>
<pre><code>usage:  umi_dedupl [--] [--help] [--opts OPTS] [--cutsites CUTSITES]
        [--platform PLATFORM] [--cutoff CUTOFF] [--emax EMAX]
        [--eperc EPERC] [--num-proc NUM-PROC] [--suffix SUFFIX]
        dirpath experiment condition

Deduplicate UMIs.

positional arguments:
  dirpath           Experiment condition directory, contains UMI counts.
  experiment        Experiment ID (e.g., TK26).
  condition         Condition folder name (e.g., 400U2h).

flags:
  -h, --help        show this help message and exit

optional arguments:
  -x, --opts OPTS           RDS file containing argument values
  -cs, --cutsites CUTSITES  Binary flag for cutsite assignment.
                            1 for, and 0 for no, cutsites
                            [default: 0]
  -p, --platform PLATFORM   Sequencing platform identifier.
                            [default: L]
  -co, --cutoff CUTOFF      Probability cutoff,
                            compared to the automatic for filtering.
                            [default: 1]
  -em, --emax EMAX          Maximum error probability for filtering.
                            [default: 0.001]
  -ep, --eperc EPERC        Maximum percentage of bases
                            with emax error probability. [default: 20]
  -c, --num-proc NUM-PROC   Number of cores for parallel computation.
                            [default: 1]
  -s, --suffix SUFFIX       Suffix to be added to output files.
                            [default: '']
</code></pre>
<p>First, the length of the UMIs is checked for inconsistencies.</p>
<p>Then, if a manual automatic <code>cutoff</code> higher than 0 was specified, a self-match probability filter is applied. Also, a single-base quality filter is applied that discards any read containing more than <code>eperc</code>% bases with error probability higher than or equal to <code>emax</code>.</p>
<p>Finally, a strict unique operation is performed that removes duplciated UMIs merely based on their sequence.</p>
<p>A figure report is generated (<code>umi_dedup.report.png</code>), showing the intensity of the detected UMI duplication.</p>
<p>The output consists of a single file, either <code>UMIpos.unique.txt</code> or <code>UMIpos.unique.atcs.txt</code>. <code>atcs</code> stands for <strong>AT CutSite</strong>. In other words, the <code>pos2cutsite</code> step is not mandatory and, if the list of custite in the current reference genome is not known, the analysis can be performed anyway.</p>
<p>The output text files contains the following columns:</p>
<ul>
<li>chromosome</li>
<li>position</li>
<li>space-separated UMI sequences</li>
</ul>
<h3 id="cs_bin"><a name="user-content-cs_bin" href="#cs_bin" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./cs_bin</h3>
<p>This R script is run by <code>main</code> in step 7, and performs the cutsite binning. The output is a <code>cs_per_bin.bsize.bstep.RData</code> file containing the binned cutsite.</p>
<pre><code>usage:  cs_bin [--] [--help] [--opts OPTS] [--bin-size BIN-SIZE]
        [--bin-step BIN-STEP] [--num-proc NUM-PROC] [--suffix SUFFIX]
        dirpath cutsites chrlengths

Bin cutsites.

positional arguments:
  dirpath           Directory to which the binned cutsites will be saved.
  cutsites          File containing the cutsite positions. (chr | pos)
  chrlengths        File containing the chromosome lengths. (chr | len)

flags:
  -h, --help        show this help message and exit

optional arguments:
  -x, --opts OPTS           RDS file containing argument values
  -i, --bin-size BIN-SIZE   Bin size in bp.
                            [default: 1e+06]
  -t, --bin-step BIN-STEP   Distance between the starting point
                            of consecutive bins in bp.
                            [default: 1e+05]
  -c, --num-proc NUM-PROC   Number of cores for parallel computation.
                            [default: 1]
  -s, --suffix SUFFIX       Suffix to be added to output files.
                            [default: '']
</code></pre>
<h3 id="umi_bin"><a name="user-content-umi_bin" href="#umi_bin" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./umi_bin</h3>
<p>This R script is run by <code>main</code> in step 7, and performs the UMI binning. The output is a <code>experiment.umi_table.bsize.bstep.RData</code> file containing the binned cutsite. Also, a <code>experiment.condition.umi_table.bsize.bstep.tsv</code> file is saved in every condition folder.</p>
<pre><code>usage:  umi_bin [--] [--help] [--opts OPTS] [--bin-size BIN-SIZE]
        [--bin-step BIN-STEP] [--num-proc NUM-PROC] [--suffix SUFFIX]
        dirpath experiment conditions cutsites chrlengths

Bin UMIs.

positional arguments:
  dirpath           Experiment main directory,
                    contains binned cutsites if available.
  experiment        Experiment ID (e.g., TK26).
  conditions        Comma-separated conditions (e.g., 400U2h,200U1h).
  cutsites          File containing the cutsite positions. (chr | pos)
  chrlengths        File containing the chromosome lengths. (chr | len)

flags:
  -h, --help        show this help message and exit

optional arguments:
  -x, --opts OPTS           RDS file containing argument values
  -i, --bin-size BIN-SIZE   Bin size in bp. [default: 1e+06]
  -t, --bin-step BIN-STEP   Distance between the starting point
                            of consecutive bins in bp.
                            [default: 1e+05]
  -c, --num-proc NUM-PROC   Number of cores for parallel computation.
                            [default: 1]
  -s, --suffix SUFFIX       Suffix to be added to output files.
                            [default: '']
</code></pre>
<h3 id="umi_plot"><a name="user-content-umi_plot" href="#umi_plot" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>./umi_plot</h3>
<p>This R script is executed by <code>main</code> in step 8. It generates the final plots of the UMI analysis. The actual plot code is stored in <code>umi_plot.functions</code>, which is imported and required by the script.</p>
<pre><code>usage:  umi_dedupl [--] [--help] [--rmChrX] [--rmChrY] [--opts OPTS]
        [--neg NEG] [--bin-size BIN-SIZE] [--bin-step BIN-STEP]
        [--num-proc NUM-PROC] [--suffix SUFFIX]
        dirpath experiment conditions cutsites chrlengths mask

Deduplicate UMIs.

positional arguments:
  dirpath       Experiment condition directory, contains UMI counts.
  experiment    Experiment ID (e.g., TK26).
  conditions    Comma-separated conditions (e.g., 400U2h,200U1h).
  cutsites      File containing the cutsite positions. (chr|pos)
  chrlengths    File containing the chromosome lengths. (chr|len)
  mask          File containing the masked regions.
                (num|chr|start|end|type)

flags:
  -h, --help        show this help message and exit
  -r, --rmChrX      Remove ChrX from the analysis
  --rmChrY          Remove ChrY from the analysis

optional arguments:
  -x, --opts OPTS           RDS file containing argument values
  -n, --neg NEG             Negative condition label.
                            [default: '']
  -i, --bin-size BIN-SIZE   Bin size in bp.
                            [default: 1e+06]
  -t, --bin-step BIN-STEP   Distance between the starting point
                            of consecutive bins in bp.
                            [default: 1e+05]
  -c, --num-proc NUM-PROC   Number of cores for parallel computation.
                            [default: 1]
  -s, --suffix SUFFIX       Suffix to be added to output files.
                            [default: '']
</code></pre>
<p>Generated plots:</p>
<ul>
<li>UMI distribution across condition.<ul>
<li>Saved as <code>experiment.umi_distribution.suff.pdf</code>.</li>
</ul>
</li>
<li>Sensed cutsites profile.<ul>
<li>Saved as <code>experiment.sensed_cutsites.bsize.bstep.pdf</code>.</li>
</ul>
</li>
<li>UMI count profiles.<ul>
<li>Plots the absolute count (<code>umi_absolute</code>), mean (<code>umi_mean</code>), standard deviation (<code>umi_sd</code>), CV (<code>umi_cv</code>) and FF (<code>umi_FF</code>).</li>
</ul>
</li>
<li>UMI probability profiles.<ul>
<li>Plots the absolute probability (<code>p_umi_absolute</code>), mean (<code>p_umi_mean</code>), standard deviation (<code>p_umi_sd</code>), CV (<code>p_umi_cv</code>) and FF (<code>p_umi_FF</code>).</li>
<li>If a negative condition is present, the <code>p_umi_mean</code> plot is generated also without the negative condition as <code>p_umi_mean.no_neg</code></li>
</ul>
</li>
<li>Difference profiles.<ul>
<li>Plots the FF and CV difference between consecutive conditions (<code>deltaFF</code> and <code>deltaCV</code>) and between the first and last conditions (<code>deltaFF.extr</code> and <code>deltaCV.extr</code>).</li>
</ul>
</li>
<li>Chromosome wide plots.<ul>
<li>Plots the average FF and CV on chromosome-wide windows, as <code>chr_wide_ff</code> and <code>chr_wide_cv</code>.</li>
</ul>
</li>
<li>Genomic overview.<ul>
<li>Produces a genome overview of the read location (binned), normalized either per chromosome (<code>genome_view.chromo</code>) or globally per condition (<code>genome_view.global</code>).</li>
</ul>
</li>
</ul></article></body></html>