# Summary

`gpseq-seq-gg` creates a summary table, containing information on the status of the nuber of reads remaining after different filter steps. New columns are added by different steps of the pipeline, specifically:

### Pattern filtering

- **condition**: condition name specified in the `pat_files`
- **pattern**: pattern used to identify the condition reads
- **total_read_count**: total number of reads in the library
- **reads_with_prefix**: number of reads matching the provided pattern (i.e., prefix)
- **prefix/total**: fraction or percentage of reads matching the provided pattern

### Read alignment

The columns added by this step are obtained through `samtools flagstats` and are not completely informative. They are replaced by *ad hoc* columns in the next step.

- **mapped**: number of generated alignments
- **mapped/prefix**: **WRONG**, please ignore this field
- **properly_paired**: number of properly paired fragments (for pair-end sequencing)
- **properly_paired/prefix**: fractiion or percentage of properly paired fragments (for pair end sequencing)
- **inter_chr**: number of fragment pairs mapped to different chromosomes (i.e., chimeras)

### Alignment filtering

The columns added by this step replace those generated in the *read alignment* step.

- **secondary_aln**: number of secondary alignments (can be ignored). Can be subtracted from the `mapped` column generated by the previous step to calculate the number of mapped reads (they are discarded)
- **chimeras**: number of read pairs mapped to different chromosomes (only for pair end sequencing, they are discarded)
- **unmapped**: number of unmapped reads
- **r2**: number of R2 fragments (only for pair end sequencing, they are discarded)
- **mapq < X**: number of alignments with mapping quality lower than the specified threshold (they are discarded)
- **rmChr**: number of reads mapped to a chromosome that is not present in the original sample (they are discarded)
- **umis**: number of remaining reads after alignment filtering
- **umis/prefix**: fraction or percentage of remaining reads after alignment filtering

### UMI deduplication

- **unmasked**: number of aligned reads that are not masked out
- **unmasked/umis**: fraction or percentage of aligned reads that are not masked out
- **assigned_to_cs**: number of unmasked reads that are assigned to a known cutsite
- **to_cs/unmasked**: fraction or percentage of unmasked reads that are assigned to a known cutsite
- **pass_read_qc**: number of reads assigned to a known cutsite that pass the read quality filter
- **read_qc/to_cs**: fraction or percentage of reads assigned to a known cutsite that pass the read quality filter
- **unique_umis**: number of de-duplicated reads that pass all the filtering steps
- **unique_read/read_qc**: fraction or percentage of deduplicated reads that pass all the previous filtering

