<!DOCTYPE html><html><head><meta charset="utf-8"><style>@font-face {
  font-family: octicons-anchor;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAYcAA0AAAAACjQAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABMAAAABwAAAAca8vGTk9TLzIAAAFMAAAARAAAAFZG1VHVY21hcAAAAZAAAAA+AAABQgAP9AdjdnQgAAAB0AAAAAQAAAAEACICiGdhc3AAAAHUAAAACAAAAAj//wADZ2x5ZgAAAdwAAADRAAABEKyikaNoZWFkAAACsAAAAC0AAAA2AtXoA2hoZWEAAALgAAAAHAAAACQHngNFaG10eAAAAvwAAAAQAAAAEAwAACJsb2NhAAADDAAAAAoAAAAKALIAVG1heHAAAAMYAAAAHwAAACABEAB2bmFtZQAAAzgAAALBAAAFu3I9x/Nwb3N0AAAF/AAAAB0AAAAvaoFvbwAAAAEAAAAAzBdyYwAAAADP2IQvAAAAAM/bz7t4nGNgZGFgnMDAysDB1Ml0hoGBoR9CM75mMGLkYGBgYmBlZsAKAtJcUxgcPsR8iGF2+O/AEMPsznAYKMwIkgMA5REMOXicY2BgYGaAYBkGRgYQsAHyGMF8FgYFIM0ChED+h5j//yEk/3KoSgZGNgYYk4GRCUgwMaACRoZhDwCs7QgGAAAAIgKIAAAAAf//AAJ4nHWMMQrCQBBF/0zWrCCIKUQsTDCL2EXMohYGSSmorScInsRGL2DOYJe0Ntp7BK+gJ1BxF1stZvjz/v8DRghQzEc4kIgKwiAppcA9LtzKLSkdNhKFY3HF4lK69ExKslx7Xa+vPRVS43G98vG1DnkDMIBUgFN0MDXflU8tbaZOUkXUH0+U27RoRpOIyCKjbMCVejwypzJJG4jIwb43rfl6wbwanocrJm9XFYfskuVC5K/TPyczNU7b84CXcbxks1Un6H6tLH9vf2LRnn8Ax7A5WQAAAHicY2BkYGAA4teL1+yI57f5ysDNwgAC529f0kOmWRiYVgEpDgYmEA8AUzEKsQAAAHicY2BkYGB2+O/AEMPCAAJAkpEBFbAAADgKAe0EAAAiAAAAAAQAAAAEAAAAAAAAKgAqACoAiAAAeJxjYGRgYGBhsGFgYgABEMkFhAwM/xn0QAIAD6YBhwB4nI1Ty07cMBS9QwKlQapQW3VXySvEqDCZGbGaHULiIQ1FKgjWMxknMfLEke2A+IJu+wntrt/QbVf9gG75jK577Lg8K1qQPCfnnnt8fX1NRC/pmjrk/zprC+8D7tBy9DHgBXoWfQ44Av8t4Bj4Z8CLtBL9CniJluPXASf0Lm4CXqFX8Q84dOLnMB17N4c7tBo1AS/Qi+hTwBH4rwHHwN8DXqQ30XXAS7QaLwSc0Gn8NuAVWou/gFmnjLrEaEh9GmDdDGgL3B4JsrRPDU2hTOiMSuJUIdKQQayiAth69r6akSSFqIJuA19TrzCIaY8sIoxyrNIrL//pw7A2iMygkX5vDj+G+kuoLdX4GlGK/8Lnlz6/h9MpmoO9rafrz7ILXEHHaAx95s9lsI7AHNMBWEZHULnfAXwG9/ZqdzLI08iuwRloXE8kfhXYAvE23+23DU3t626rbs8/8adv+9DWknsHp3E17oCf+Z48rvEQNZ78paYM38qfk3v/u3l3u3GXN2Dmvmvpf1Srwk3pB/VSsp512bA/GG5i2WJ7wu430yQ5K3nFGiOqgtmSB5pJVSizwaacmUZzZhXLlZTq8qGGFY2YcSkqbth6aW1tRmlaCFs2016m5qn36SbJrqosG4uMV4aP2PHBmB3tjtmgN2izkGQyLWprekbIntJFing32a5rKWCN/SdSoga45EJykyQ7asZvHQ8PTm6cslIpwyeyjbVltNikc2HTR7YKh9LBl9DADC0U/jLcBZDKrMhUBfQBvXRzLtFtjU9eNHKin0x5InTqb8lNpfKv1s1xHzTXRqgKzek/mb7nB8RZTCDhGEX3kK/8Q75AmUM/eLkfA+0Hi908Kx4eNsMgudg5GLdRD7a84npi+YxNr5i5KIbW5izXas7cHXIMAau1OueZhfj+cOcP3P8MNIWLyYOBuxL6DRylJ4cAAAB4nGNgYoAALjDJyIAOWMCiTIxMLDmZedkABtIBygAAAA==) format('woff');
}

* {
    box-sizing: border-box;
}

body {
    width: 980px;
    margin-right: auto;
    margin-left: auto;
}

body .markdown-body {
    padding: 45px;
    border: 1px solid #ddd;
    border-radius: 3px;
    word-wrap: break-word;
}

pre {
    font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}

.markdown-body {
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  color: #333;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 16px;
  line-height: 1.6;
  word-wrap: break-word;
}

.markdown-body a {
  background-color: transparent;
}

.markdown-body a:active,
.markdown-body a:hover {
  outline: 0;
}

.markdown-body strong {
  font-weight: bold;
}

.markdown-body h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

.markdown-body img {
  border: 0;
}

.markdown-body hr {
  box-sizing: content-box;
  height: 0;
}

.markdown-body pre {
  overflow: auto;
}

.markdown-body code,
.markdown-body kbd,
.markdown-body pre {
  font-family: monospace, monospace;
  font-size: 1em;
}

.markdown-body input {
  color: inherit;
  font: inherit;
  margin: 0;
}

.markdown-body html input[disabled] {
  cursor: default;
}

.markdown-body input {
  line-height: normal;
}

.markdown-body input[type="checkbox"] {
  box-sizing: border-box;
  padding: 0;
}

.markdown-body table {
  border-collapse: collapse;
  border-spacing: 0;
}

.markdown-body td,
.markdown-body th {
  padding: 0;
}

.markdown-body input {
  font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}

.markdown-body a {
  color: #4078c0;
  text-decoration: none;
}

.markdown-body a:hover,
.markdown-body a:active {
  text-decoration: underline;
}

.markdown-body hr {
  height: 0;
  margin: 15px 0;
  overflow: hidden;
  background: transparent;
  border: 0;
  border-bottom: 1px solid #ddd;
}

.markdown-body hr:before {
  display: table;
  content: "";
}

.markdown-body hr:after {
  display: table;
  clear: both;
  content: "";
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 15px;
  margin-bottom: 15px;
  line-height: 1.1;
}

.markdown-body h1 {
  font-size: 30px;
}

.markdown-body h2 {
  font-size: 21px;
}

.markdown-body h3 {
  font-size: 16px;
}

.markdown-body h4 {
  font-size: 14px;
}

.markdown-body h5 {
  font-size: 12px;
}

.markdown-body h6 {
  font-size: 11px;
}

.markdown-body blockquote {
  margin: 0;
}

.markdown-body ul,
.markdown-body ol {
  padding: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body ol ol,
.markdown-body ul ol {
  list-style-type: lower-roman;
}

.markdown-body ul ul ol,
.markdown-body ul ol ol,
.markdown-body ol ul ol,
.markdown-body ol ol ol {
  list-style-type: lower-alpha;
}

.markdown-body dd {
  margin-left: 0;
}

.markdown-body code {
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
}

.markdown-body pre {
  margin-top: 0;
  margin-bottom: 0;
  font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}

.markdown-body .select::-ms-expand {
  opacity: 0;
}

.markdown-body .octicon {
  font: normal normal normal 16px/1 octicons-anchor;
  display: inline-block;
  text-decoration: none;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.markdown-body .octicon-link:before {
  content: '\f05c';
}

.markdown-body:before {
  display: table;
  content: "";
}

.markdown-body:after {
  display: table;
  clear: both;
  content: "";
}

.markdown-body>*:first-child {
  margin-top: 0 !important;
}

.markdown-body>*:last-child {
  margin-bottom: 0 !important;
}

.markdown-body a:not([href]) {
  color: inherit;
  text-decoration: none;
}

.markdown-body .anchor {
  display: inline-block;
  padding-right: 2px;
  margin-left: -18px;
}

.markdown-body .anchor:focus {
  outline: none;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 1em;
  margin-bottom: 16px;
  font-weight: bold;
  line-height: 1.4;
}

.markdown-body h1 .octicon-link,
.markdown-body h2 .octicon-link,
.markdown-body h3 .octicon-link,
.markdown-body h4 .octicon-link,
.markdown-body h5 .octicon-link,
.markdown-body h6 .octicon-link {
  color: #000;
  vertical-align: middle;
  visibility: hidden;
}

.markdown-body h1:hover .anchor,
.markdown-body h2:hover .anchor,
.markdown-body h3:hover .anchor,
.markdown-body h4:hover .anchor,
.markdown-body h5:hover .anchor,
.markdown-body h6:hover .anchor {
  text-decoration: none;
}

.markdown-body h1:hover .anchor .octicon-link,
.markdown-body h2:hover .anchor .octicon-link,
.markdown-body h3:hover .anchor .octicon-link,
.markdown-body h4:hover .anchor .octicon-link,
.markdown-body h5:hover .anchor .octicon-link,
.markdown-body h6:hover .anchor .octicon-link {
  visibility: visible;
}

.markdown-body h1 {
  padding-bottom: 0.3em;
  font-size: 2.25em;
  line-height: 1.2;
  border-bottom: 1px solid #eee;
}

.markdown-body h1 .anchor {
  line-height: 1;
}

.markdown-body h2 {
  padding-bottom: 0.3em;
  font-size: 1.75em;
  line-height: 1.225;
  border-bottom: 1px solid #eee;
}

.markdown-body h2 .anchor {
  line-height: 1;
}

.markdown-body h3 {
  font-size: 1.5em;
  line-height: 1.43;
}

.markdown-body h3 .anchor {
  line-height: 1.2;
}

.markdown-body h4 {
  font-size: 1.25em;
}

.markdown-body h4 .anchor {
  line-height: 1.2;
}

.markdown-body h5 {
  font-size: 1em;
}

.markdown-body h5 .anchor {
  line-height: 1.1;
}

.markdown-body h6 {
  font-size: 1em;
  color: #777;
}

.markdown-body h6 .anchor {
  line-height: 1.1;
}

.markdown-body p,
.markdown-body blockquote,
.markdown-body ul,
.markdown-body ol,
.markdown-body dl,
.markdown-body table,
.markdown-body pre {
  margin-top: 0;
  margin-bottom: 16px;
}

.markdown-body hr {
  height: 4px;
  padding: 0;
  margin: 16px 0;
  background-color: #e7e7e7;
  border: 0 none;
}

.markdown-body ul,
.markdown-body ol {
  padding-left: 2em;
}

.markdown-body ul ul,
.markdown-body ul ol,
.markdown-body ol ol,
.markdown-body ol ul {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body li>p {
  margin-top: 16px;
}

.markdown-body dl {
  padding: 0;
}

.markdown-body dl dt {
  padding: 0;
  margin-top: 16px;
  font-size: 1em;
  font-style: italic;
  font-weight: bold;
}

.markdown-body dl dd {
  padding: 0 16px;
  margin-bottom: 16px;
}

.markdown-body blockquote {
  padding: 0 15px;
  color: #777;
  border-left: 4px solid #ddd;
}

.markdown-body blockquote>:first-child {
  margin-top: 0;
}

.markdown-body blockquote>:last-child {
  margin-bottom: 0;
}

.markdown-body table {
  display: block;
  width: 100%;
  overflow: auto;
  word-break: normal;
  word-break: keep-all;
}

.markdown-body table th {
  font-weight: bold;
}

.markdown-body table th,
.markdown-body table td {
  padding: 6px 13px;
  border: 1px solid #ddd;
}

.markdown-body table tr {
  background-color: #fff;
  border-top: 1px solid #ccc;
}

.markdown-body table tr:nth-child(2n) {
  background-color: #f8f8f8;
}

.markdown-body img {
  max-width: 100%;
  box-sizing: content-box;
  background-color: #fff;
}

.markdown-body code {
  padding: 0;
  padding-top: 0.2em;
  padding-bottom: 0.2em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(0,0,0,0.04);
  border-radius: 3px;
}

.markdown-body code:before,
.markdown-body code:after {
  letter-spacing: -0.2em;
  content: "\00a0";
}

.markdown-body pre>code {
  padding: 0;
  margin: 0;
  font-size: 100%;
  word-break: normal;
  white-space: pre;
  background: transparent;
  border: 0;
}

.markdown-body .highlight {
  margin-bottom: 16px;
}

.markdown-body .highlight pre,
.markdown-body pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f7f7f7;
  border-radius: 3px;
}

.markdown-body .highlight pre {
  margin-bottom: 0;
  word-break: normal;
}

.markdown-body pre {
  word-wrap: normal;
}

.markdown-body pre code {
  display: inline;
  max-width: initial;
  padding: 0;
  margin: 0;
  overflow: initial;
  line-height: inherit;
  word-wrap: normal;
  background-color: transparent;
  border: 0;
}

.markdown-body pre code:before,
.markdown-body pre code:after {
  content: normal;
}

.markdown-body kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb;
}

.markdown-body .pl-c {
  color: #969896;
}

.markdown-body .pl-c1,
.markdown-body .pl-s .pl-v {
  color: #0086b3;
}

.markdown-body .pl-e,
.markdown-body .pl-en {
  color: #795da3;
}

.markdown-body .pl-s .pl-s1,
.markdown-body .pl-smi {
  color: #333;
}

.markdown-body .pl-ent {
  color: #63a35c;
}

.markdown-body .pl-k {
  color: #a71d5d;
}

.markdown-body .pl-pds,
.markdown-body .pl-s,
.markdown-body .pl-s .pl-pse .pl-s1,
.markdown-body .pl-sr,
.markdown-body .pl-sr .pl-cce,
.markdown-body .pl-sr .pl-sra,
.markdown-body .pl-sr .pl-sre {
  color: #183691;
}

.markdown-body .pl-v {
  color: #ed6a43;
}

.markdown-body .pl-id {
  color: #b52a1d;
}

.markdown-body .pl-ii {
  background-color: #b52a1d;
  color: #f8f8f8;
}

.markdown-body .pl-sr .pl-cce {
  color: #63a35c;
  font-weight: bold;
}

.markdown-body .pl-ml {
  color: #693a17;
}

.markdown-body .pl-mh,
.markdown-body .pl-mh .pl-en,
.markdown-body .pl-ms {
  color: #1d3e81;
  font-weight: bold;
}

.markdown-body .pl-mq {
  color: #008080;
}

.markdown-body .pl-mi {
  color: #333;
  font-style: italic;
}

.markdown-body .pl-mb {
  color: #333;
  font-weight: bold;
}

.markdown-body .pl-md {
  background-color: #ffecec;
  color: #bd2c00;
}

.markdown-body .pl-mi1 {
  background-color: #eaffea;
  color: #55a532;
}

.markdown-body .pl-mdr {
  color: #795da3;
  font-weight: bold;
}

.markdown-body .pl-mo {
  color: #1d3e81;
}

.markdown-body kbd {
  display: inline-block;
  padding: 3px 5px;
  font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb;
}

.markdown-body .plan-price-unit {
  color: #767676;
  font-weight: normal;
}

.markdown-body .task-list-item {
  list-style-type: none;
}

.markdown-body .task-list-item+.task-list-item {
  margin-top: 3px;
}

.markdown-body .task-list-item input {
  margin: 0 0.35em 0.25em -1.6em;
  vertical-align: middle;
}

.markdown-body .plan-choice {
  padding: 15px;
  padding-left: 40px;
  display: block;
  border: 1px solid #e0e0e0;
  position: relative;
  font-weight: normal;
  background-color: #fafafa;
}

.markdown-body .plan-choice.open {
  background-color: #fff;
}

.markdown-body .plan-choice.open .plan-choice-seat-breakdown {
  display: block;
}

.markdown-body .plan-choice-free {
  border-radius: 3px 3px 0 0;
}

.markdown-body .plan-choice-paid {
  border-radius: 0 0 3px 3px;
  border-top: 0;
  margin-bottom: 20px;
}

.markdown-body .plan-choice-radio {
  position: absolute;
  left: 15px;
  top: 18px;
}

.markdown-body .plan-choice-exp {
  color: #999;
  font-size: 12px;
  margin-top: 5px;
}

.markdown-body .plan-choice-seat-breakdown {
  margin-top: 10px;
  display: none;
}

.markdown-body :checked+.radio-label {
  z-index: 1;
  position: relative;
  border-color: #4078c0;
}
</style><title>README</title></head><body><article class="markdown-body"><h1>
<a id="user-content-genomic-loci-positioning-by-sequencing-gpseq-sequencing-data-analysis" class="anchor" href="#genomic-loci-positioning-by-sequencing-gpseq-sequencing-data-analysis" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Genomic loci Positioning by Sequencing (GPSeq) sequencing data analysis</h1>

<h2>
<a id="user-content-the-main-script" class="anchor" href="#the-main-script" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>The <code>main</code> script</h2>

<pre><code>usage: ./main.sh [-h][-w][-t threads] -i inDir -o outDir -e expID -c conditions
 [-n][-a aligner][-g refGenome][-d bwaIndex][-x][-y][-f cutsite][-q mapqThr]
 [-p platform][-u umiLength][-r csRange][-j emax][-k eperc][-z binSize]
 [-b binStep][-l csList][-m maskFile][-s chrLengths]

 Description:
  Run a step-by-step interactive GPSeq sequencing data analysis.

 Required files:
  Requires R1 (and R2 if paired-end sequencing) and a pattern files in the
  input directory (inDir). The patterns file should have a condition per row
  with condition name, pattern (scan_for_mateches format) and pattern length
  separated by tabulations.

 Mandatory arguments:
  -i indir      Input directory.
  -o outdir     Output directory. Created if not found.
  -e expID      Experiment ID.
  -c conditions Comma-separated conditions.

 Optional arguments:
  -h            Show this help page.
  -y            Perform every step of the pipeline without asking.
  -n            Negative present. Expected label: neg;
  -x            Remove X chromosome after alignment.
  -y            Remove Y chromosome after alignment.
  -t threads    Number of threads for parallelization.
  -a aligner    Aligner. Either 'bwa' (default) or 'bowtie2'.
  -g refGenome  Path to reference genome file. Default: 'hg19'.
  -d bwaIndex   Path to BWA index file. Required if BWA is the selected aligner.
  -f cutsite    Cutsite sequence. Default: 'AAGCTT' (HindIII).
  -q mapqThr    Mapping quality threshold. Default: 30.
  -p platform   Sequencing platform. Default: 'L'.
  -u umilength  UMI sequence length. Default: 8.
  -r csRange    Range around cutsite for UMI assignment. Default: 40.
  -j emax       Maximum error probability for read quality filtering. Default: 1e-3.
  -k eperc      Maximum % of bases with emax error probability. Default: 20.
  -z binSize    Bin size. Default: 1e6.
  -b binStep    Bin step. Default: 1e5.
  -l csList     File with cutsite list. Columns: chr|pos. No header.
  -m maskFile   File with masked regions. Columns: id|chr|start|end. No header.
  -s chrLengths File with chromosome lengths. chr|len. No header.
</code></pre>

<h4>
<a id="user-content-pipeline-steps" class="anchor" href="#pipeline-steps" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Pipeline steps</h4>

<p>The steps performed by <code>main</code> are the following:</p>

<ol>
<li>Run <code>./quality_control.sh</code>.</li>
<li>Run <code>./files_prepare.sh</code>
</li>
<li>Perform pattern filtering:

<ul>
<li>Count total reads and prepare <code>summary</code> file</li>
<li>Run <code>./pattern_filtering.sh</code> per condition and update the <code>summary</code> file</li>
</ul>
</li>
<li>Perform read alignment.

<ul>
<li>First, run <code>./reads_trim.sh</code> trim the pattern from the R1 reads (length specified in <code>pat_files</code>)</li>
<li>Run <code>./reads_align.sh</code> and update summary with fraction of mapped reads</li>
<li>Add the patterns to the SAM files (condition.linker.sam)</li>
</ul>
</li>
<li>Filters the SAM file. (this step requires a specified <code>mapqthr</code> setting)

<ul>
<li>Run <code>./sam_filter</code>.</li>
</ul>
</li>
<li>Prepare UMIs. If a <code>maskfile</code> was specified, UMIs are masked based on the information contained in such file.

<ul>
<li>Run <code>./umi_group</code> to group UMIs on the same position.</li>
<li>Run <code>./pos2cutsite</code> to group UMIs on a cutsite</li>
<li>Run <code>./umi_dedupl</code> to deduplicate UMIs.</li>
</ul>
</li>
<li>Performs binning.

<ul>
<li>Run <code>./cs_bin</code> to bin the cutsites.</li>
<li>Run <code>./umi_bin</code> to bin the UMIs.</li>
</ul>
</li>
<li>Analyzes UMIs.

<ul>
<li>Run <code>./umi_plot</code> to generate analysis results as figures.</li>
</ul>
</li>
</ol>

<h4>
<a id="user-content-output" class="anchor" href="#output" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Output</h4>

<p>Description of different output files.</p>

<h4>
<a id="user-content-input" class="anchor" href="#input" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input</h4>

<p>At least two files are required: a sequencing platform output file (R1), and a <code>pat_files</code> file in the input directory, which contains the pattern to recognize the different conditions. If the sequencing is paired ended, also an R2 file is required.</p>

<p>The <code>pat_files</code>, which is expected to be located in the input folder, contains a row per condition (as specified in the <code>-c conditions</code> argument). Every row contains the following tab-separated information: condition label, scan_for_matches pattern, pattern length. An example being the following:</p>

<pre><code>neg ^ 8...8 CATCAGAA AAGCTT 1...1000 $  22
1min    ^ 8...8 CATCATCC AAGCTT 1...1000 $  22
10min   ^ 8...8 GTCGTTCC AAGCTT 1...1000 $  22
2h  ^ 8...8 TGATGTCC AAGCTT 1...1000 $  22
</code></pre>

<h2>
<a id="user-content-other-scripts" class="anchor" href="#other-scripts" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Other scripts</h2>

<h3>
<a id="user-content-quality_controlsh" class="anchor" href="#quality_controlsh" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./quality_control.sh</h3>

<p>This script is run by <code>main</code> during step 1. It performs a quality control with the <code>fastqc</code> tool.</p>

<pre><code> usage: ./quality_control.sh [-h][-t threads] -o outDir -1 r1file [-2 r2file]

 Description:
  Run FASTQC tool.

 Mandatory arguments:
  -o outDir Output folder, created if not found.
  -1 r1file R1 fastq file.

 Optional arguments:
  -h        Show this help page.
  -t threads    Number of threads for parallelization
  -2 r2file R2 fastq file.
</code></pre>

<h3>
<a id="user-content-files_preparesh" class="anchor" href="#files_preparesh" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./files_prepare.sh</h3>

<p>This script is run by <code>main</code> during step 2. It decompresses the sequencing data and converts the generated files in different 'formats'.</p>

<pre><code> usage: ./files_prepare.sh [-h][-t threads] -o outDir -1 r1file [-2 r2file]

 Description:
  Run FASTQC tool.

 Mandatory arguments:
  -o outDir Output folder, created if not found.
  -1 r1file R1 fastq file.

 Optional arguments:
  -h        Show this help page.
  -t threads    Number of threads for parallelization
  -2 r2file R2 fastq file.
</code></pre>

<p>The produced files are the following:</p>

<ul>
<li>
<code>r1.fq</code> (and <code>r2.fq</code>): decompressed <strong>fastq</strong> file(s).</li>
<li>
<code>r2oneline.fq</code> (and <code>r2oneline.fq</code>): one read per line <strong>fastq</strong> file(s).</li>
<li>
<code>r1.fa</code> (and <code>r2.fa</code>): decompressed <strong>fasta</strong> file(s).</li>
<li>
<code>r1oneline.fa</code> (and <code>r2oneline.fa</code>): one read per line <strong>fasta</strong> file(s).</li>
</ul>

<h3>
<a id="user-content-pattern_filter" class="anchor" href="#pattern_filter" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./pattern_filter</h3>

<p>This script is run by <code>main</code> during step 3. It runs the <code>scan_for_matches</code> tool to select reads based on a pattern (e.g., containing barcodes), and divide them into different conditions.</p>

<pre><code> usage: ./pattern_filter.sh [-h][-t threads] -i inDir -o outDir -p patFile

 Description:
  Select reads based on the provided pattern by runnin scan_for_matches.

 Mandatory arguments:
  -i inDir  Input directory.
  -o outDir Output directory.
  -p patFile    Pattern file.

 Optional arguments:
  -h        Show this help page.
  -t threads    Number of threads for parallelization.
</code></pre>

<p>The script will create a <code>filtered.r1.fa</code> and a <code>filtered.r1.fq</code> (and respective r2 files in case of paired-end sequencing) in the specified <code>outDir</code>. These files will contain the reads that match the specified pattern, retrieved from <code>r1.fa</code> and <code>r1.fq</code> (and r2...).</p>

<p>Finally, the <code>summary</code> file is updated with the number and percentage of reads belonging to that condition.</p>

<h3>
<a id="user-content-reads_trimsh" class="anchor" href="#reads_trimsh" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./reads_trim.sh</h3>

<p>This BASH script is run by <code>main</code> in step 4. It removes and stores the linkers from the R1 reads to a separate file.</p>

<pre><code> usage: ./reads_trim.sh [-h][-t threads] -o outDir -c cond -p patFile

 Description:
  Trim linker from reads.

 Mandatory arguments:
  -o outDir     Output directory. Created if not found.
  -c cond       Condition to analyze.
  -p patFile    Pattern file.

 Optional arguments:
  -h            Show this help page.
  -t threads    Number of threads for parallelization.
</code></pre>

<h3>
<a id="user-content-reads_alignsh" class="anchor" href="#reads_alignsh" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./reads_align.sh</h3>

<p>This BASH script is run by <code>main</code> in step 4. It aligns the reads to a reference genome using either <code>bwa</code> or <code>bowtie2</code> and generates <code>bam</code> and <code>bai</code> files.</p>

<pre><code> usage: ./reads_align.sh [-h][-t threads] -o outDir -c cond
        [-p][-r ref][-a aln]

 Description:
  Align reads using either bowtie2 or bwa mem.

 Mandatory arguments:
  -o outDir     Output directory. Created if not found.
  -c cond       Condition to analyze.

 Optional arguments:
  -h            Show this help page.
  -t threads    Number of threads for parallelization.
  -p            Option for paired-end sequencing.
  -r ref        Reference ref. Default: hg19.
  -a aln        Aligner. Either 'bwa' or 'bowtie2'. Default: 'bwa'.
  -i index      Path to BWA index.
                Default: '$DATA'/BiCro-Resources/genomes/'$ref'bwt/'$ref'.fa
</code></pre>

<h3>
<a id="user-content-sam_filterr" class="anchor" href="#sam_filterr" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./sam_filter.R</h3>

<p>This R script is run by <code>main</code> in step 5. It filters a SAM file based on a variety of settings.</p>

<pre><code>usage:  sam_filter.R [--help][--cutsite CUTSITE][--mapq-thr MAPQ-THR]
        [--num-proc NUM-PROC][--suffix SUFFIX] dirpath experiment condition

Filter alignment output.

positional arguments:
  dirpath           Experiment ccondition directory, contains UMI counts.
  experiment        Experiment ID (e.g., TK26).
  condition         Condition folder name (e.g., 400U2h).

flags:
  -h, --help        show this help message and exit

optional arguments:
  -cs, --cutsite CUTSITE        Cutsite sequence (e.g., for HindIII AAGCTT).
                                [default: AAGCTT]
  -mt, --mapq-thr MAPQ-THR      Mapping quality threshold.
                                [default: 30]
  -c, --num-proc NUM-PROC       Number of cores for parallel computation.
                                [default: 1]
  -s, --suffix SUFFIX           Suffix to be added to output files.
                                [default: '']
</code></pre>

<p>Specifically, the filters do the following:</p>

<ul>
<li>Remove secondary alignments (based on SAM file flags).</li>
<li>If paired-end sequencing, remove chimeras (where R1 and R2 are mapped on different chromosomes).</li>
<li>Remove unmapped reads.</li>
<li>If paired-end sequencing, remove R2 and keep only R1.</li>
<li>Remove reads with a mapping quality lower than <code>mapqthr</code>.</li>
</ul>

<p>Then, the generated output consists of two files:</p>

<ul>
<li>
<code>condition.filtered.sam.RData</code> contains the raw filtered SAM file as an RData structure.</li>
<li>
<code>condition.filtered.umi.pos.txt</code> is a table with the following columns: chromosome, 3'position, linker sequence, linker quality, read name. The position does not correspond to the SAM file one, but is instead shifted accordingly to the CIGAR string in the SAM file or to the length of the cutsite.</li>
</ul>

<h3>
<a id="user-content-umi_grouppy" class="anchor" href="#umi_grouppy" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./umi_group.py</h3>

<p>This Python script is run by <code>main</code> in step 6. It groups reads mapped on the same genomic locus.</p>

<pre><code>usage: umi_group.py [-h] [--mask-file MF] dirpath expID cond umiLen

Group UMIs based on their location.

positional arguments:
  dirpath         Experiment condition directory, contains UMI counts.
  expID           Experiment ID (e.g., TK26).
  cond            Condition folder name (e.g., 400U2h).
  umiLen          Length of the UMIs in nt (e.g., 8).

optional arguments:
  -h, --help      show this help message and exit
  --mask-file MF  File with regions to be masked.
</code></pre>

<p>The output consists of:</p>

<ul>
<li>
<code>UMIpos.txt</code>, containing the following columns:

<ul>
<li>chromosome</li>
<li>position</li>
<li>space-separated UMIs</li>
<li>quality strings</li>
</ul>
</li>
<li>If a <code>maskfile</code> was specified, then an additional <code>UMIpos.all.txt</code> file will be generated, containing all the UMIs (also those withing masked regions).</li>
</ul>

<h3>
<a id="user-content-pos2cutsiter" class="anchor" href="#pos2cutsiter" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./pos2cutsite.R</h3>

<p>This R script is run by <code>main</code> in step 6. It groups reads mapped on the same cutsite. Every read is assigned to the closest cutsite. Then, a read is discarded and considered 'orphan', if the closest cutsite is more than bin_size/2 nt away from the aligned locus. In other words, a window of bin_size nt is built around every cutsite.</p>

<p>The <code>pos2cutsite</code> step is not mandatory and, if the list of custite in the current reference genome is not known, the analysis can be performed anyway.</p>

<pre><code>usage:  pos2cutsite.R [--help][--bin-size BIN-SIZE][--num-proc NUM-PROC]
        [--suffix SUFFIX] dirpath experiment condition cutsites

Convert genomic coordinate to cutsite.

positional arguments:
  dirpath           Experiment ccondition directory, contains UMI counts.
  experiment        Experiment ID (e.g., TK26).
  condition         Condition folder name (e.g., 400U2h).
  cutsites          File containing the cutsite positions. (chr | pos)

flags:
  -h, --help        show this help message and exit

optional arguments:
  -i, --bin-size BIN-SIZE   Range size in bp around the cutsite.
                            [default: 40]
  -c, --num-proc NUM-PROC   Number of cores for parallel computation.
                            [default: 1]
  -s, --suffix SUFFIX       Suffix to be added to output files.
                            [default: '']
</code></pre>

<h3>
<a id="user-content-umi_deduplr" class="anchor" href="#umi_deduplr" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./umi_dedupl.R</h3>

<p>This R script is run by <code>main</code> in step 6. It removes duplicated (through amplification) reads based on the UMIs.</p>

<pre><code>usage:  umi_dedupl.R [--help][--cutsites CUTSITES][--platform PLATFORM]
        [--cutoff CUTOFF][--emax EMAX][--eperc EPERC][--num-proc NUM-PROC]
        [--suffix SUFFIX] dirpath experiment condition

Deduplicate UMIs.

positional arguments:
  dirpath           Experiment condition directory, contains UMI counts.
  experiment        Experiment ID (e.g., TK26).
  condition         Condition folder name (e.g., 400U2h).

flags:
  -h, --help        show this help message and exit

optional arguments:
  -cs, --cutsites CUTSITES  Binary flag for cutsite assignment.
                            1 for, and 0 for no, cutsites
                            [default: 0]
  -p, --platform PLATFORM   Sequencing platform identifier.
                            [default: L]
  -co, --cutoff CUTOFF      Probability cutoff,
                            compared to the automatic for filtering.
                            [default: 1]
  -em, --emax EMAX          Maximum error probability for filtering.
                            [default: 0.001]
  -ep, --eperc EPERC        Maximum percentage of bases
                            with emax error probability. [default: 20]
  -c, --num-proc NUM-PROC   Number of cores for parallel computation.
                            [default: 1]
  -s, --suffix SUFFIX       Suffix to be added to output files.
                            [default: '']
</code></pre>

<p>First, the length of the UMIs is checked for inconsistencies.</p>

<p>Then, if a manual automatic <code>cutoff</code> higher than 0 was specified, a self-match probability filter is applied. Also, a single-base quality filter is applied that discards any read containing more than <code>eperc</code>% bases with error probability higher than or equal to <code>emax</code>.</p>

<p>Finally, a strict unique operation is performed that removes duplciated UMIs merely based on their sequence.</p>

<p>A figure report is generated (<code>umi_dedup.report.png</code>), showing the intensity of the detected UMI duplication.</p>

<p>The output consists of a single file, either <code>UMIpos.unique.txt</code> or <code>UMIpos.unique.atcs.txt</code>. <code>atcs</code> stands for <strong>AT CutSite</strong>. In other words, the <code>pos2cutsite</code> step is not mandatory and, if the list of custite in the current reference genome is not known, the analysis can be performed anyway.</p>

<p>The output text files contains the following columns:</p>

<ul>
<li>chromosome</li>
<li>position</li>
<li>space-separated UMI sequences</li>
</ul>

<h3>
<a id="user-content-cs_binr" class="anchor" href="#cs_binr" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./cs_bin.R</h3>

<p>This R script is run by <code>main</code> in step 7, and performs the cutsite binning. The output is a <code>cs_per_bin.bsize.bstep.RData</code> file containing the binned cutsite.</p>

<pre><code>usage:  cs_bin.R [--help][-i BIN-SIZE][-t BIN-STEP]
        [-c NUM-PROC][-s SUFFIX] dirpath cutsites chrlengths

Bin cutsites.

positional arguments:
  dirpath                     Directory to which the binned cutsites will be saved.
  cutsites                    File containing the cutsite positions. (chr | pos)
  chrlengths                  File containing the chromosome lengths. (chr | len)

flags:
  -h, --help                  show this help message and exit

optional arguments:
  -i, --bin-size BIN-SIZE     Bin size in bp.
                              [default: 1e+06]
  -t, --bin-step BIN-STEP     Distance between the starting point of consecutive bins in bp.
                              [default: 1e+05]
  -c, --num-proc NUM-PROC     Number of cores for parallel computation.
                              [default: 1]
  -s, --suffix SUFFIX         Suffix to be added to output files.
                              [default: ]
</code></pre>

<h3>
<a id="user-content-umi_binr" class="anchor" href="#umi_binr" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./umi_bin.R</h3>

<p>This R script is run by <code>main</code> in step 7, and performs the UMI binning. The output is a <code>experiment.umi_table.bsize.bstep.RData</code> file containing the binned cutsite. Also, a <code>experiment.condition.umi_table.bsize.bstep.tsv</code> file is saved in every condition folder.</p>

<pre><code>usage:  umi_bin.R [-h][-c][-i BIN-SIZE][-t BIN-STEP][-p NUM-PROC]
        [-s SUFFIX] dirpath experiment conditions chrlengths

Bin UMIs.

positional arguments:
  dirpath         Experiment main directory, contains binned cutsites if available.
  experiment      Experiment ID (e.g., TK26).
  conditions      Comma-separated conditions (e.g., 400U2h,200U1h).
  chrlengths      File containing the chromosome lengths. (chr | len)

flags:
  -h, --help           show this help message and exit
  -c, --cutsites       Whether cutsites where used.

optional arguments:
  -i, --bin-size BIN-SIZE     Bin size in bp.
                              [default: 1e+06]
  -t, --bin-step BIN-STEP     Distance between the starting point of consecutive bins in bp.
                              [default: 1e+05]
  -p, --num-proc NUM-PROC     Number of cores for parallel computation.
                              [default: 1]
  -s, --suffix SUFFIX         Suffix to be added to output files.
                              [default: ]

</code></pre>

<h3>
<a id="user-content-umi_plotr" class="anchor" href="#umi_plotr" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>./umi_plot.R</h3>

<p>This R script is executed by <code>main</code> in step 8. It generates the final plots of the UMI analysis. The actual plot code is stored in <code>umi_plot.functions</code>, which is imported and required by the script.</p>

<pre><code>usage:  umi_plot.R [--] [--help] [--rmChrX] [--rmChrY] [--opts OPTS]
    [--neg NEG] [--bin-size BIN-SIZE] [--bin-step BIN-STEP]
    [--num-proc NUM-PROC] [--suffix SUFFIX]
    dirpath experiment conditions cutsites chrlengths mask

Generate UMI plots.

positional arguments:
dirpath       Experiment condition directory, contains UMI counts.
experiment    Experiment ID (e.g., TK26).
conditions    Comma-separated conditions (e.g., 400U2h,200U1h).
cutsites      File containing the cutsite positions. (chr|pos)
chrlengths    File containing the chromosome lengths. (chr|len)
mask          File containing the masked regions.
            (num|chr|start|end|type)

flags:
-h, --help        show this help message and exit
-r, --rmChrX      Remove ChrX from the analysis
--rmChrY          Remove ChrY from the analysis

optional arguments:
-x, --opts OPTS           RDS file containing argument values
-n, --neg NEG             Negative condition label.
                        [default: '']
-i, --bin-size BIN-SIZE   Bin size in bp.
                        [default: 1e+06]
-t, --bin-step BIN-STEP   Distance between the starting point
                        of consecutive bins in bp.
                        [default: 1e+05]
-c, --num-proc NUM-PROC   Number of cores for parallel computation.
                        [default: 1]
-s, --suffix SUFFIX       Suffix to be added to output files.
                        [default: '']
</code></pre>

<p>Generated plots:</p>

<ul>
<li>UMI distribution across condition.

<ul>
<li>Saved as <code>experiment.umi_distribution.suff.pdf</code>.</li>
</ul>
</li>
<li>Sensed cutsites profile.

<ul>
<li>Saved as <code>experiment.sensed_cutsites.bsize.bstep.pdf</code>.</li>
</ul>
</li>
<li>UMI count profiles.

<ul>
<li>Plots the absolute count (<code>umi_absolute</code>), mean (<code>umi_mean</code>), standard deviation (<code>umi_sd</code>), CV (<code>umi_cv</code>) and FF (<code>umi_FF</code>).</li>
</ul>
</li>
<li>UMI probability profiles.

<ul>
<li>Plots the absolute probability (<code>p_umi_absolute</code>), mean (<code>p_umi_mean</code>), standard deviation (<code>p_umi_sd</code>), CV (<code>p_umi_cv</code>) and FF (<code>p_umi_FF</code>).</li>
<li>If a negative condition is present, the <code>p_umi_mean</code> plot is generated also without the negative condition as <code>p_umi_mean.no_neg</code>
</li>
</ul>
</li>
<li>Difference profiles.

<ul>
<li>Plots the FF and CV difference between consecutive conditions (<code>deltaFF</code> and <code>deltaCV</code>) and between the first and last conditions (<code>deltaFF.extr</code> and <code>deltaCV.extr</code>).</li>
</ul>
</li>
<li>Chromosome wide plots.

<ul>
<li>Plots the average FF and CV on chromosome-wide windows, as <code>chr_wide_ff</code> and <code>chr_wide_cv</code>.</li>
</ul>
</li>
<li>Genomic overview.

<ul>
<li>Produces a genome overview of the read location (binned), normalized either per chromosome (<code>genome_view.chromo</code>) or globally per condition (<code>genome_view.global</code>).</li>
</ul>
</li>
</ul>
</article></body></html>